code coverage foundation for hash and num_counters

richkadel opened the pull request:
This PR is the next iteration after PR #73011 (which is still waiting on bors to merge).

@wesleywiser - PTAL
r? @tmandry 

(FYI, I'm also working on injecting the coverage maps, in another branch, while waiting for these to merge.)

Thanks!

Commit:
code coverage foundation for hash and num_counters

Replaced dummy values for hash and num_counters with computed values,
and refactored InstrumentCoverage pass to simplify injecting more
counters per function in upcoming versions.

Improved usage documentation and error messaging.

Commit:
Merge pull request #4 from rust-lang/master

update from origin 2020-06-18

richkadel commented:
FYI, this PR (currently) represents only 1 commit since #73011 but until that one lands, the changes in both PRs show up in "Files changed". To see only the diffs in this PR, use this link:

https://github.com/rust-lang/rust/pull/73488/commits/4caf8823107ad17ed2dd8cafe634429fc6c986e6

wesleywiser commented on src/librustc_codegen_llvm/intrinsic.rs:
```diff
@@ -141,26 +141,17 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
-                } else {
-                    bug!("intrinsic count_code_region: no src.instance");
-                }
+                let coverage_data = fx.mir.coverage_data.as_ref().unwrap();
```
I think instead of directly accessing the MIR body here, we should add a new query to `TyCtxt` that can return the coverage data for given `DefId`. Then you won't need to take the `FunctionCx` here or make `FunctionCx.mir` or `FunctionCx.instance` public or add the `coverage_data` field to `Body`. 

That integrates better with our incremental system and wouldn't require these other changes.

richkadel replied:
Great idea. Thanks!

richkadel replied:
@wesleywiser I'm a little unsure what you are suggesting for the query, particularly in the context of _not_ storing the `coverage_data` in the `mir::Body`.

I'm trying to familiarize myself with the queries and provider implementations, but I don't know how to set it up yet.

If I implement a `query coverage_data(DefId) -> mir::CoverageData`, but create the `CoverageData` during the `InstrumentCoverage` pass, where do I save the `CoverageData` so that the query implementation will return it?

It seems like it might be possible to store it directly in something like the rustc_metadata::rmeta tables, such as:

```rust
    coverage_data: Table<DefIndex, Lazy<mir::CoverageData>>,
```

or maybe add a new Map/cache directly to `GlobalCtxt`?

But there are probably other options, and in all of these cases, I can't pinpoint any examples in MIR-time processing where data like this is computed, stored, and then retrieved from a `query` downstream, without invoking the query to compute the value.

Any suggestions or code references I can model this after?

Thanks!
Rich

wesleywiser replied:
So, what I was envisioning was a two-step approach where

- the `InstrumentCoverage ` MIR pass would only inject counters and nothing else

- the `coverage_data` query would get the MIR, count how many counters are in the MIR and hash the HIR body

The result of the `coverage_data` query would then be automatically persisted for you in the compiler session context. If you ever need that data cross-crate in the future, this would also set you up to be able to do that quite easily.

richkadel replied:
FYI, I was uploading a new commit just as you sent your reply.

Got it. I understand how that would work in the current context, but I wonder if this will have the best performance in the future.

I'm thinking of 2 or 3 possible upcoming changes that might still make it less efficient to split computing the state data (in the query) from instrumenting the MIR (in the pass). For example:

1. Just re-counting the injected calls requires walking the MIR again.
2. The implementation of the hash may change (Note that clang's approach does some coverage-specific analysis, so the hash is less sensitive to changes that don't affect flow).

With that in mind, is it reasonable to stash the coverage data in the `Body` struct, or do you feel strongly that this state data should be computed separately from the coverage analysis?

richkadel replied:
(Note, the new commit addresses part of your recommendation; it does implement the `query`, so I don't need to change the `codegen_intrinsic_call()` function signature, and can now retrieve the CoverageData from `tcx.coverage_data(def_id)`, which is an improvement. But the query gets the CoverageData from the `Body`.)

richkadel replied:
@wesleywiser - @tmandry  and I discussed this. We're good with either the current approach to implementing the query, or counting the intrinsic calls by traversing the MIR in the query as per your last suggestion. Once we know your preference, I believe Tyler is ready to approve.

wesleywiser replied:
Thanks for being flexible about this @richkadel! I talked a bit about this with @oli-obk and we think for now not storing the `CoverageData` in the `Body` and instead re-walking the MIR is the better architecture. If we discover that causes performance issues, we'd certainly be open to a different approach to improve performance but at this time, we don't think there will be much of a performance hit and it integrated better with the rest of the compiler.

richkadel replied:
Will do. Thanks!

richkadel replied:
Done. PTAL. Thanks!

wesleywiser replied:
This looks great to me. Thank you!

bors commented:
:umbrella: The latest upstream changes (presumably #73504) made this pull request unmergeable. Please resolve the merge conflicts.

Commit:
Merge pull request #5 from rust-lang/master

update from origin 2020-06-19

richkadel commented on src/librustc_middle/mir/mod.rs:
```diff
@@ -88,6 +88,19 @@ impl MirPhase {
+    pub hash: u64,
+
+    /// The total number of coverage region counters added to this MIR Body.
+    pub num_counters: usize,
```
Agreed. I tried u32 first, but for some reason u64 and usize are Foldable, and u32 is not; and I couldn't figure out where that's getting set. I'll look more closely.

richkadel replied:
I'm able to get it to work by explicitly implementing `TypeFoldable` for `u32`. Will upload for review shortly.

richkadel replied:
Oh, looking at your last comment more closely, if I don't have to add coverage data to the body, I may not need it to be FoldableType. Hmm...

wesleywiser replied:
Seems fine to add an impl for it (I think this is the correct place)

https://github.com/rust-lang/rust/blob/a39c7787ba246353178e099373b9240be0d9e603/src/librustc_middle/ty/structural_impls.rs#L260

but if we can get away without it, that's even better.

richkadel replied:
FYI, I had put it here:

https://github.com/rust-lang/rust/blob/a39c7787ba246353178e099373b9240be0d9e603/src/librustc_middle/mir/type_foldable.rs#L322

I'm not sure it matters, but since it's a primitive (not related to `ty`) I thought keeping the impl with the type definition made sense. It's hard to tell why some impls are where they are for this trait.

richkadel replied:
I went ahead and committed these changes while I'm waiting for any additional guidance you can provide on your other suggested change to implement coverage_data as a query.

Commit:
num_counters to u32, after implementing TypeFoldable

Commit:
implemented query for coverage data

This commit adds a query that allows the CoverageData to be pulled from
a call on tcx, avoiding the need to change the
`codegen_intrinsic_call()` signature (no need to pass in the FunctionCx
or any additional arguments.

The commit does not change where/when the CoverageData is computed. It's
still done in the `pass`, and saved in the MIR `Body`.

See discussion (in progress) here:
https://github.com/rust-lang/rust/pull/73488#discussion_r443825646

richkadel commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -79,14 +126,59 @@ fn function_handle<'tcx>(tcx: TyCtxt<'tcx>, fn_def_id: DefId, span: Span) -> Ope
+    stable_hasher.finish()
+}
+
+fn associated_body<'hir>(node: Node<'hir>) -> Option<BodyId> {
```
FYI, this was copied verbatim from:

https://github.com/richkadel/rust/blob/754da8849c18c45ff2fd2e77213f8371de488f80/src/librustc_middle/hir/map/mod.rs#L59

If there's no objection, I'd like to make the original function public and just use it directly.

tmandry replied:
Making it public seems fine. I'm not sure it should (continue to) live there, but I'm also not sure where it should go instead.

richkadel replied:
done

tmandry commented on src/librustc_metadata/locator.rs:
```diff
@@ -488,6 +488,8 @@ impl<'a> CrateLocator<'a> {
             {
                 err.note(&format!("the `{}` target may not be installed", self.triple));
+            } else if self.crate_name == sym::profiler_builtins {
+                err.note(&"the compiler may have been built without `profiler = true`");
```
nit: let's not reference specific compiler configuration keys in an error message, I think "without profiler runtimes" (or something that maps to its comment in config.toml.example) should suffice.

richkadel replied:
done

tmandry commented on src/librustc_middle/query/mod.rs:
```diff
@@ -214,6 +214,12 @@ rustc_queries! {
         }
 
+        query coverage_data(key: DefId) -> Option<mir::CoverageData> {
+            desc { |tcx| "retrieving coverage data, if computed from MIR for `{}`", tcx.def_path_str(key) }
```
nit: if this description is showing up in our backtrace, it's very likely that the "if computed" clause is true, so I'd suggest taking it out for clarity

richkadel replied:
done

Commit:
Address remaining feedback items

wesleywiser commented on src/librustc_middle/query/mod.rs:
```diff
@@ -214,6 +214,12 @@ rustc_queries! {
             cache_on_disk_if { key.is_local() }
         }
 
+        query coverage_data(key: DefId) -> Option<mir::CoverageData> {
```
When you make this query operate by walking the `Body` MIR, I think it would be reasonable to change the return type to just `mir::CoverageData`. The query should be able to operate on any `Body` even ones that aren't instrumented since it will just return zero `num_counters`.

richkadel replied:
Makes sense to me. Note, `num_counters == 0` implies there are no calls to `instrprof.increment`, and therefore no need for the `hash` argument for the given function, so I will also set the hash to `0` in that case.

richkadel replied:
Done. PTAL. Thanks!

Commit:
moves coverage data computation from pass to query

Commit:
PR no longer requires u32 impl TypeFoldable

Commit:
using "mir_body" (vs "body") in InstrumentCoverage

The mod uses both MIR bodies and HIR bodies, so I'm trying to maintain
consistency with these names.

tmandry commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -19,6 +21,31 @@ use rustc_span::Span;
+                {
+                    if let FnDef(called_fn_def_id, _) = func.literal.ty.kind {
+                        if called_fn_def_id == count_code_region_fn {
+                            num_counters += 1;
```
It occurs to me that a MIR optimization could (a) duplicate a counter, or (b) inline counters from other functions. How should we deal with cases like that?

For (a) we could keep a set of counter id's we've already seen to deduplicate.

For handling (b), I'm assuming we'll want the name of the function where the counter originated in source? We could pass a def_id of the source function as part of the "intrinsic". But then how would we keep track of monomorphizations? We can always address this point in a follow-up PR, but it's worth thinking about.

richkadel replied:
For (a), I'll change the query implementation to return the max(counters) + 1. This fits well with the LLVM spec for instrprof.increment.

But (b) raises some interesting questions. Assuming it is possible for the MIR to include BasicBlocks from multiple "functions" in the source (which sounds logical), then I'm thinking the same thing you suggested; I'll probably need to add the DefId of the function in the injected Call terminator, rather than assuming the code region to be counted is part of the function represented by the MIR's DefId.

richkadel replied:
I addressed (a) in the latest commit.

For (b) I added FIXME comments for now. I'd like to address this in a separate PR, but will start looking into it.

richkadel replied:
Oh, I just remembered, we did talk about this (effect of inlining, for example) and my recommendation at the time is still something I'd like to propose:  Disable inlining, and any other optimizations that might "break" code coverage, if the instrument_coverage option is turned on.

I'm not sure if we can address all of the edge cases like (b) with this, but I think it's plausible.

WDYT?

tmandry replied:
We can disable inlining in that case, sure (it's still not enabled by default). It seems like we might want a solution for this long-term, but it doesn't seem like a super high priority.

Commit:
Updated query for num_counters to compute from max index

Also added FIXME comments to note the possible need to accommodate
counter increment calls in source-based functions that differ from the
function context of the caller instance (e.g., inline functions).

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit 977ce57d915914139c4aa643e63f368913e5f437 has been approved by `tmandry`

<!-- @bors r=tmandry 977ce57d915914139c4aa643e63f368913e5f437 -->
<!-- homu: {"type":"Approved","sha":"977ce57d915914139c4aa643e63f368913e5f437","approver":"tmandry"} -->
